telegram/ext/_extbot.py

```diff
@@ -121,12 +212,101 @@ def __init__(
             self.arbitrary_callback_data = arbitrary_callback_data
         self.callback_data_cache: CallbackDataCache = CallbackDataCache(bot=self, maxsize=maxsize)
 
+    async def initialize(self) -> None:
+        # Initialize before calling super, because super calls get_me
+        if self.rate_limiter:
+            await self.rate_limiter.initialize()
+        await super().initialize()
+
+    async def shutdown(self) -> None:
+        # Shut down the rate limiter before shutting down the request objects!
+        if self.rate_limiter:
+            await self.rate_limiter.shutdown()
+        await super().shutdown()
+
+    @classmethod
+    def _merge_api_rl_kwargs(
+        cls, api_kwargs: Optional[JSONDict], rate_limit_args: Optional[RLARGS]
+    ) -> Optional[JSONDict]:
+        """Inserts the `rate_limit_args` into `api_kwargs` with the special key `__RL_KEY` so
+        that we can extract them later without having to modify the `telegram.Bot` class.
+        """
+        if not rate_limit_args:
+            return api_kwargs
+        if api_kwargs is None:
+            api_kwargs = {}
+        api_kwargs[cls.__RL_KEY] = rate_limit_args
+        return api_kwargs
+
+    @classmethod
+    def _extract_rl_kwargs(cls, data: Optional[JSONDict]) -> Optional[RLARGS]:
+        """Extracts the `rate_limit_args` from `data` if it exists."""
+        if not data:
+            return None
+        return data.pop(cls.__RL_KEY, None)
+
+    async def _do_post(
+        self,
+        endpoint: str,
+        data: JSONDict,
+        *,
+        read_timeout: ODVInput[float] = DEFAULT_NONE,
+        write_timeout: ODVInput[float] = DEFAULT_NONE,
+        connect_timeout: ODVInput[float] = DEFAULT_NONE,
+        pool_timeout: ODVInput[float] = DEFAULT_NONE,
+    ) -> Union[bool, JSONDict, None]:
+        rate_limit_args = self._extract_rl_kwargs(data)
+        if not self.rate_limiter and rate_limit_args is not None:
+            raise ValueError(
+                "`rate_limit_args` can only be used if a `ExtBot.rate_limiter` is set."
+            )
+
+        # getting updates should not be rate limited!
+        if endpoint == "getUpdates" or not self.rate_limiter:
+            return await super()._do_post(
```

Comment:
shouldn't we call `self.get_updates` here? Hmm, that would lead to infinite recursion.. or I guess it's not supposed to be used here?

Reply:
Sorry, what's not supposed to be used here? `_post` and `_do_post` are the methods that are called by every bot method internally, including `getUpdates`. Here we just make sure that `getUpdates` isn't rate limited

Reply:
nvm, I got a bit confused, all good now